### PR TITLE
Feat: string utility library

### LINKED ISSUE #158: [FEAT]: Create a basic string utility library
## 📘 Description
Create a basic string utility library in Cairo focusing on the most essential string operations. This initial version will serve as a foundation for more complex string manipulations in future iterations.

## ✅ Acceptance Criteria
Implement basic string functions:
- concatenate: join two strings together
- length: get string length
- starts_with: check if string starts with a prefix
- ends_with: check if string ends with a suffix

Include:
- Basic test cases for each function
- Simple usage examples

## 🌎 References
- [Cairo String Documentation](https://book.cairo-lang.org/)

## 📜 Additional Notes
- ⚠ Focus on basic functionality first
- ⚠ Keep implementations simple and readable

### PR DESCRIPTION
**Pull request by armaanansari121**
# 📝 Feat: String utility library

## 🛠️ Issue
- Closes #158 

## 📖 Description
Implemented the string library with the following components.
- `String` struct which represents the String that we are working with.
- `new` method to create an empty string.
- `len` method to retrieve the length of the string.
- `concatenate` method to safely concatenate two strings, This method takes a reference to the string to which another string is to be concatenated.
- `starts_with` method to check if a string starts with a substring.
- `ends_with` method to check if a string ends with a substring.

## 🖼️ Screenshots - Tests
image

## 📝 Additional Notes
- Added integration tests for each function considering proper edge cases.
- Added some real world use cases in the tests. 
- Added proper documentation for each function to improve code maintainability.

### CODE REVIEW

**Review by coxmars — CHANGES_REQUESTED** [review]
@armaanansari121 Great job, it looks pretty good to me. Also, can you create a .tool-versions please 🫡

**Comment by armaanansari121** [comment]
> @armaanansari121 Great job, it looks pretty good to me. Also, can you create a .tool-versions please 🫡

Sure give me a minute.

**Comment by armaanansari121** [comment]
The changes have been made, ser.

**Comment by armaanansari121** [comment]
@coxmars seems like the crowdfunding contract has some failing tests. Please let me know if I need to change anything.

**Review by coxmars — APPROVED** [review]
LGTM!

File: examples/cairo/scripts/string_utility/tests/test_contract.cairo [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+#[test]
+fn test_len() {
+    // Test empty string
+    let empty_string = StringTrait::new();
+    assert(empty_string.len() == 0, 'empty string length failed');
+
+    // Test non-empty string
+    let data: ByteArray = "Hello";
+    let string = String { data };
```
Comment (danielcdz) [inline examples/cairo/scripts/string_utility/tests/test_contract.cairo:17]: To better demonstrate the library's functionality and eliminate the need for `ByteArray`, what do you think about directly initializing the string variable within the struct? 

Instead of first declaring a `ByteArray`, we could write something like this:  

```rust
let string = String { "Hello" };
```  

If it is possible then refactor all the necessary use cases

Comment (armaanansari121) [reply examples/cairo/scripts/string_utility/tests/test_contract.cairo:17]: This was the initial thought, however we get the error that data field is not provided.
Best we can do is
```
let string = String { data: "Hello" };
```

There's another way to better do this, which is to change the new function to take in a parameter so we can then initialise new strings in the following way.
```
let string = StringTrait::new("Hello"); 
```
I personally think this is a better way.

Comment (armaanansari121) [reply examples/cairo/scripts/string_utility/tests/test_contract.cairo:17]: I just made the changes locally using the `StringTrait::new("Hello")` syntax and it looks pretty clean. 
Let me know what you think and I'll commit the changes.

Comment (danielcdz) [reply examples/cairo/scripts/string_utility/tests/test_contract.cairo:17]: @armaanansari121 go ahead!

**Comment by armaanansari121** [comment]
Done ser! @danielcdz

**Review by danielcdz — APPROVED** [review]

**Comment by danielcdz** [comment]
@armaanansari121 looks pretty clean, thanks for your collaboration!
